Lg/srtool update (#1467)
substrate-developer-hub/substrate-docs: content/md/en/docs/reference/command-line-tools/srtool.md

---
title: srtool
description: Describes the Docker container and command-line interface for building the Substrate WebAssembly runtime.
keywords:
---

> The Substrate runtime tool (`srtool`) is a Docker container that enables you to build the Substrate WebAssembly runtime in a deterministic way.

Reviewer: It may be worth mentioning that `srtool` started as a docker image but grew as a set of tools around the docker image.

Author: good suggestion.

By using this tool, you can ensure that the same source code consistently reproduces an identical WebAssembly blob. 
You can also use the tool to inspect and audit the runtime for any Substrate-based chain and to integrate building the WebAssembly runtime into your CI/CD pipeline.

## Get the Docker image

You must have a Docker account and Docker installed in your build environment to access the Substrate runtime tool container.

> If you sign in to the Docker Hub, you can search for the `paritytech/srtool` container and corresponding tagged images that are available there.

Reviewer: Not sure we should start with the Docker image. It is rather easy to make mistakes.
The `srtool-cli` has been made to make things MUCH easier. I would keep this doc, but slide the docker details down for the people who want to go deeper.

Author: Yeah, good point. I documented using Docker first because this is literally what I did first -- before finding that the CLI makes this easier than going through Docker. Live and learn ....

### Image naming convention

> The naming convention for `paritytech/srtool` Docker images specifies the version of the Rust compiler used to compile the code included in the image.

Reviewer: ```suggestion

IMPORTANT: Unlike many Docker image, `srtool` there is no `latest` tag for `srtool` and you must knowingly provide the compiler version as described below.

The naming convention for `paritytech/srtool` Docker images specifies the version of the Rust compiler used to compile the code included in the image.
```

For example, the image `paritytech/srtool:1.62.0` indicates that the tools included in the image were compiled with version `1.62.0` of the `rustc` compiler.
In addition to the compiler version, some images specify the version of the build script used.
For example, an image named `paritytech/srtool:1.62.0-0.9.19` was compiled with version `1.62.0` of the `rustc` compiler but using the version `0.9.19` of the build script.

> The image that only specifies the compiler version always contains the latest version of the software.

Reviewer: ```suggestion
NOTE: The image that only specifies the compiler version always contains the latest version of the software.
```

### Download an image from the Docker Hub

You can pull a  `paritytech/srtool` container image directly from Docker Hub or install and use one of the [helper programs](#get-helper-programs) for working with the  `paritytech/srtool` container.

To pull the image from the Docker Hub:

1. Sign in to Docker Hub.

2. Type `paritytech/srtool` in the Search field and press Enter.

3. Click **paritytech/srtool**, then click **Tags**.

4. Copy the command for the image you want to pull.
   
5. Open a terminal shell on your local computer.
   
6. Paste the command you copied from the Docker Hub.
   
   For example, you should run a command similar to the following:

   ```bash
   docker pull paritytech/srtool:1.62.0
   ```

>    The command downloads and unpacks the image.

Reviewer: This is precisely the reason why I recommend using `srtool-cli`, the whole thing you describe can then be done as:
```
srtool pull
```

## Get helper programs

> The Substrate runtime tool requires Docker.

Reviewer: ```suggestion
`srtool` requires Docker.
```

However, you don't need to know anything about using Docker to build a chain using `srtool`.
Instead, you can use the following programs to interact with the Docker image.

- [srtool-cli](https://github.com/chevdor/srtool-cli) provides a command-line interface to build the runtime using the `srtool` Docker image.

> - [srtool-app](https://gitlab.com/chevdor/srtool-app) provides a simple graphical user interface for building the runtime using the `srtool` Docker image.
> 
> - [srtool-actions](https://github.com/chevdor/srtool-actions) provides Github actions to integrate builds produced using the `srtool` image with your CI/CD pipeline.
> 
> - [subwasm](https://github.com/chevdor/subwasm) provides command-line options for working with the metadata and WebAssembly runtime built using `srtool`.

Reviewer: In order of usefulness :)

```suggestion
- [subwasm](https://github.com/chevdor/subwasm) provides command-line options for working with the metadata and WebAssembly runtime built using `srtool`.

- [srtool-actions](https://github.com/chevdor/srtool-actions) provides Github actions to integrate builds produced using the `srtool` image with your CI/CD pipeline.

- [srtool-app](https://gitlab.com/chevdor/srtool-app) provides a simple graphical user interface for building the runtime using the `srtool` Docker image.
```

| `-n`, `--no-cache` | Specifies that you don't want to use the tag value from the local cache. |
| `-V`, `--version` | Displays version information.   |

### Subcommands

You can use the following subcommands with the `srtool` command.

| Command | Description |
| ------- | ----------- |
| `build` | Starts a new `srtool` container to build your runtime. |
| `help` | Displays usage information for `srtool` or for a specified subcommand. |
| `info` | Displays information about the `srtool` container and your repository. | 
| `pull` | Pulls the `srtool` image without running anything else. |
| `version` | Displays version information for the `srtool` container. Use `--version` if you want version information for the `srtool-cli` executable.

### Examples

To get version information for the `srtool` Docker image, run the following command:

```bash
srtool version
```

The command displays output similar to the following:

```text
{
  "name": "srtool",
  "version": "0.9.21",
  "rustc": "1.62.0",
  "subwasm": "0.18.0",
  "tera": "0.2.1",
  "toml": "0.2.1"
}

> ```

Reviewer: Beware, if you show a run of each sub command, this may quick become deprecated.
I do that in the main repo because I use Asciidoc and I have scripts to automatically produce the output and generate the doc. Here I would simply put a link towards the repo, that will simplify the maintenance for you.

Author: Warning taken. I think I'll leave the examples in for now (since I was able to walk through all of them and it is generally nice to provide sampleinput and resulting output), but I can remove them/replace them with links if maintenance becomes as issue.

---
title: srtool
description: Describes the Docker container and command-line interface for building the Substrate WebAssembly runtime.
keywords:
---

The core component of the Substrate runtime toolbox (`srtool`) is a Docker container that enables you to build the Substrate WebAssembly runtime in a deterministic way.
By using this tool, you can ensure that the same source code consistently reproduces an identical WebAssembly blob. 
You can also use the tool to inspect and audit the runtime for any Substrate-based chain and to integrate building the WebAssembly runtime into your CI/CD pipeline.

## Working with the Docker container

Because `srtool` is a Docker container, you must have Docker available in your build environment to use it.

> However, you don't need to know anything about using Docker to build a chain using `srtool`.

Reviewer: ```suggestion
However, you don't need to know anything about using Docker to build a chain using `srtool` thanks to the `srtool-cli`.
`srtiool-cli` is a command-line utility written in Rust that will install a utility called `srtool` on your system. It greatly simplifies your interactions with the `srtool` Docker container.
```

> Instead, you can use the following programs to interact with the Docker image.
> 
> - [srtool-cli](https://github.com/chevdor/srtool-cli) provides a command-line interface to pull the `srtool` Docker image and build the runtime using the `srtool` Docker container.
> 
> - [subwasm](https://github.com/chevdor/subwasm) provides command-line options for working with the metadata and WebAssembly runtime built using `srtool`.
> 
> - [srtool-actions](https://github.com/chevdor/srtool-actions) provides Github actions to integrate builds produced using the `srtool` image with your CI/CD pipeline.
>   
> - [srtool-app](https://gitlab.com/chevdor/srtool-app) provides a simple graphical user interface for building the runtime using the `srtool` Docker image.

Reviewer: ```suggestion
Overtime, several tools were extracted and/or joined the tooling options around `srtool` and it is worth mentioning the following tools and helpers:

- [srtool-cli](https://github.com/chevdor/srtool-cli) provides a command-line interface to pull the `srtool` Docker image, build the runtime using the `srtool` Docker container and interact with it.

- [subwasm](https://github.com/chevdor/subwasm) provides command-line options for working with the metadata and WebAssembly runtime built using `srtool`. `subwasm` is used internally as well in the `srtool` image.

- [srtool-actions](https://github.com/chevdor/srtool-actions) provides Github actions to integrate builds produced using the `srtool` image with your Github CI/CD pipelines.
  
- [srtool-app](https://gitlab.com/chevdor/srtool-app) provides a simple graphical user interface for building the runtime using the `srtool` Docker image.
```

> ## Image naming convention

Reviewer: This section could move  all the way down below the `srtool-cli`.
Also, people enabling the `srtool-cli` logger can easily see and learn about what happens under the hood.
That makes the info about Docker image naming not so relevant.

FYI: Both `srtool-cli` and `srtool-actions` have been primary made so the normal user does NOT need to know what is the canonical version of the srtiool image at the moment, those tools know how to check and fetch **the right** image/version.

Author: I've decided to split the information into two separate topics, one that has the why and how and this one focused on the command usage and examples.
